### PR TITLE
Unused Variable Hints

### PR DESCRIPTION
**Pull request by Olly-**
Added hints for unused variables (either variable, parameter or result). The added directive {$HINT} acts the same as other directives and can be toggled on and off throughout the script.

### CODE REVIEW

File: lpcompiler.pas [hunk 1]
```diff
@@ -871,8 +872,11 @@ function TLapeCompiler.HandleDirective(Sender: TLapeTokenizerBase; Directive, Ar
   begin
     Argument := LowerCase(Trim(Argument));
     if (Argument = 'on') or (Argument = '+') then
-      Include(FOptions, Option)
-    else if (Argument = 'off') or (Argument = '-') then
+    begin
+      Include(FOptions, Option);
+      if (Option = lcoHints) then
+        FHasHintDefine := True;
```
Comment (nielsAD) [inline lpcompiler.pas]: Could you elaborate on why you added `FHasHintDefine`? I don't think it's really needed.

In any case, this check should not be here. If it has to be done, it has to be done where `setOption` is called. Now, every setOption call has to check whether it's for `lcoHints` while we already know the answer when we call it.

File: lpcompiler.pas [hunk 2]
```diff
@@ -3131,6 +3156,7 @@ constructor TLapeCompiler.Create(
   FDefines.Duplicates := dupIgnore;
   FDefines.CaseSensitive := LapeCaseSensitive;
   FConditionalStack := TLapeConditionalStack.Create(0);
+  FHasHintDefine := False;
```
Comment (nielsAD) [inline lpcompiler.pas]: Defined in `TLapeCompilerBase`, so it should be initialized there as well? (if it's needed)

File: lptree.pas [hunk 3]
```diff
@@ -5156,6 +5157,23 @@ function TLapeTree_Method.Compile(var Offset: Integer): TResVar;
...
+  if (FCompiler.HasHintDefine) then
+    with FStackInfo do
+      for i := 0 to VarCount - 1 do
+      begin
+        if (Vars[i].Used <> duFalse) or (Vars[i].Name = '') or (Vars[i].Name[1] = '!') then
+          Continue;
+
+        if (Vars[i] is TLapeParameterVar) then
+        begin
+          if (FStackInfo.Vars[i].Name = 'Result') then
```
Comment (nielsAD) [inline lptree.pas]: I think this might better checking for `out` parameters, rather than only `Result`.

File: lpvartypes.pas [hunk 4]
```diff
@@ -4539,6 +4564,14 @@ function TLapeCompilerBase.hasDeclaration(ADecl: TLapeDeclaration; LocalOnly: Bo
   Result := hasDeclaration(ADecl, FStackInfo, LocalOnly, CheckWith);
 end;
 
+procedure TLapeCompilerBase.Hint(Msg: lpString; Args: array of const; ADocPos: TDocPos);
+begin
+  Assert(lcoHints in Options);
```
Comment (nielsAD) [inline lpvartypes.pas]: With the way it works now (whether hints are on or off depends on the state during definition), should this check not be removed?

File: lpvartypes.pas [hunk 5]
```diff
@@ -4539,6 +4564,14 @@ function TLapeCompilerBase.hasDeclaration(ADecl: TLapeDeclaration; LocalOnly: Bo
...
+  Assert({$IFDEF FPC}@{$ENDIF}FOnHint <> nil);
```
Comment (nielsAD) [inline lpvartypes.pas]: I think this should be turned into an `if` rather than an `Assert`.

File: lpcompiler.pas [hunk 6]
```diff
@@ -1279,7 +1287,7 @@ function TLapeCompiler.ParseMethodHeader(out Name: lpString; addToScope: Boolean
       else if (LapeCase(Name) = LapeCase(AVar)) or hasDeclaration(AVar, True) then
         LapeExceptionFmt(lpeDuplicateDeclaration, [AVar], Tokenizer.DocPos)
       else
-        FStackInfo.addVar(ParType, VarType, AVar);
+        FStackInfo.addVar(ParType, VarType, AVar, not (lcoHints in FOptions));
```
Comment (nielsAD) [inline lpcompiler.pas]: Not sure if it's worth adding a parameter for `StackInfo.addVar` for this. I think this is the only place where you need to set `.Used` to `False` / `Ignore`. You could probably get rid of the extra parameter, set `.Used` to `Ignore` by default, then set it to `False` here.

File: lpcompiler.pas [hunk 7]
```diff
@@ -1785,13 +1795,13 @@ function TLapeCompiler.ParseMethod(FuncForwards: TLapeFuncForwards; FuncHeader:
...
     begin
       if (FuncHeader is TLapeType_MethodOfType) then
-        FStackInfo.addVar(TLapeType_MethodOfType(FuncHeader).SelfParam, TLapeType_MethodOfType(FuncHeader).ObjectType, 'Self')
+        FStackInfo.addSelfVar(TLapeType_MethodOfType(FuncHeader).SelfParam, TLapeType_MethodOfType(FuncHeader).ObjectType)
       else if (FuncHeader is TLapeType_MethodOfObject) then
-        FStackInfo.addVar(lptConstRef, getGlobalType('ConstPointer'), 'Self');
+        FStackInfo.addSelfVar(lptConstRef, getGlobalType('ConstPointer'));
       for i := 0 to FuncHeader.Params.Count - 1 do
-        FStackInfo.addVar(FuncHeader.Params[i].ParType, FuncHeader.Params[i].VarType, lpString('Param'+IntToStr(i)));
+        FStackInfo.addVar(FuncHeader.Params[i].ParType, FuncHeader.Params[i].VarType, lpString('Param'+IntToStr(i)), not (lcoHints in FOptions));
```
Comment (nielsAD) [inline lpcompiler.pas]: Hints are not really important here, I think? Probably only used internally.

File: lpcompiler.pas [hunk 8]
```diff
@@ -2264,6 +2273,9 @@ function TLapeCompiler.ParseVarBlock(OneOnly: Boolean = False; ValidEnd: EParser
           VarDecl.VarDecl := addLocalVar(VarType, Identifiers[i]);
 
         VarDecl.Default := nil;
+        if (not (lcoHints in FOptions)) then
+          VarDecl.VarDecl.Used := duIgnore;
```
Comment (nielsAD) [inline lpcompiler.pas]: I think this is better than adding a parameter to `StackInfo.addVar`. But probably better to use `Ignore` by default and then switch to `False` to enable hints.

File: lpcompiler.pas [hunk 9]
```diff
@@ -2623,8 +2636,20 @@ function TLapeCompiler.ParseExpression(ReturnOn: EParserTokenSet = []; FirstNext
...
             Expr := getExpression(Tokenizer.TokString, getPDocPos());
             if (Expr = nil) then
-              LapeExceptionFmt(lpeUnknownDeclaration, [Tokenizer.TokString], Tokenizer.DocPos)
-            else if (Expr is TLapeTree_Invoke) then
+              LapeExceptionFmt(lpeUnknownDeclaration, [Tokenizer.TokString], Tokenizer.DocPos);
+
+            if (lcoHints in FOptions) then
+              Used := duTrue
+            else
+              Used := duIgnore;
```
Comment (nielsAD) [inline lpcompiler.pas]: Why set it to `Ignore` on _usage_?

File: lpvartypes.pas [hunk 10]
```diff
@@ -3522,9 +3533,21 @@ function TLapeStackInfo.addVar(VarType: TLapeType; Name: lpString = ''): TLapeSt
...
 
-function TLapeStackInfo.addVar(ParType: ELapeParameterType; VarType: TLapeType; Name: lpString = ''): TLapeStackVar;
+function TLapeStackInfo.addVar(ParType: ELapeParameterType; VarType: TLapeType; Name: lpString; IgnoreUsed: Boolean): TLapeStackVar;
+begin
+  if (VarType = nil) then
+    Result := addVar(TLapeParameterVar.Create(ParType, VarType, nil, Name))
+  else
+    Result := addVar(TLapeParameterVar.Create(ParType, VarType, nil, Name, @VarType._DocPos));
+
+  if (IgnoreUsed) then
```
Comment (nielsAD) [inline lpvartypes.pas]: As said before, not sure if it's worth doing here.

**Comment by nielsAD** [comment]
Thanks for the updates! I think it looks good except for a few details for which I added comments (most notably the `addVar` and `FHasHintDefine` changes).

**Comment by 3FLLC** [comment]
AWESOME! I implemented this, it is nice - makes LA-PE much more useful for huge projects. I agree, it needs to look out the "out" parameters. I have expanded my implementation to support "If HintCount>0 then DoNotRun"... or -C "compile only" mode. ;-)

@nielsAD where would be the best place to check "InScript" methods to trap unused ones? And then for "compile only" mode - to optimize out non-InScript methods too... like a -O1 feature.

**Comment by nielsAD** [comment]
Thanks, @Olly- !

**Pull request merged**